pilot/cmd/pilot-discovery/server/server.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// ConsulArgs provides configuration for the Consul service registry.
+type ConsulArgs struct {
+	Config    string
+	ServerURL string
+}
+
+// EurekaArgs provides configuration for the Eureka service registry
+type EurekaArgs struct {
+	ServerURL string
+}
+
+// ServiceArgs provides the composite configuration for all service registries in the system.
+type ServiceArgs struct {
+	Registries []string
+	Consul     ConsulArgs
+	Eureka     EurekaArgs
+}
+
+// PilotArgs provides all of the configuration parameters for the Pilot discovery service.
+type PilotArgs struct {
+	DiscoveryOptions envoy.DiscoveryServiceOptions
+	Namespace        string
+	Mesh             MeshArgs
+	Config           ConfigArgs
+	Service          ServiceArgs
+}
+
+// Server contains the runtime configuration for the Pilot discovery service.
+type Server struct {
+	config *pilotConfig
+}
+
+// NewServer creates a new Server instance based on the provided arguments.
+func NewServer(args PilotArgs, stop chan struct{}) (*Server, error) {
+	config := &pilotConfig{stop: stop}
+
+	// Apply the arguments to the configuration.
+	if err := aggregateConfigFn(&args, config); err != nil {
+		return nil, multierror.Prefix(err, "failed to create Pilot discovery service.")
+	}
+
+	return &Server{
+		config: config,
+	}, nil
+}
+
+// Run starts the pilot discovery service.
+func (s Server) Run() {
+	// Now start all of the services.
+	for _, runFn := range s.config.runFuncs {
+		runFn()
+	}
+}
+
+// runFunc is a no-op function that will be executed after the Pilot discovery service has been configured.
+type runFunc func()
+
+// pilotConfig contains the runtime structures for the Pilot discovery service. This structure is configured
+// by a number of pilotConfigFunc functions.
+type pilotConfig struct {
+	stop              chan struct{}
+	mesh              *meshconfig.MeshConfig
+	serviceController *aggregate.Controller
+	configController  model.ConfigStoreCache
+	mixerSAN          []string
+	runFuncs          []runFunc
+}
+
+// pilotConfigFunc is a function that configures some part of the pilotConfig based on the settings in the PilotArgs.
+type pilotConfigFunc func(in *PilotArgs, out *pilotConfig) error
+
+// initMesh creates the mesh in the pilotConfig from the input arguments.
+func initMesh(in *PilotArgs, out *pilotConfig) error {
+	// If a config file was specified, use it.
+	if in.Mesh.ConfigFile != "" {
+		mesh, err := cmd.ReadMeshConfig(in.Mesh.ConfigFile)
+		if err == nil {
+			out.mesh = mesh
+			return nil
+		}
+		glog.Warningf("failed to read mesh configuration, using default: %v", err)
+	}
+
+	// Config file either wasn't specified or failed to load - use a default mesh.
+	defaultMesh := proxy.DefaultMeshConfig()
+	mesh := &defaultMesh
+
+	// Allow some overrides for testing purposes.
+	if in.Mesh.MixerAddress != "" {
+		mesh.MixerAddress = in.Mesh.MixerAddress
+	}
+	if in.Mesh.RdsRefreshDelay != nil {
+		mesh.RdsRefreshDelay = in.Mesh.RdsRefreshDelay
+	}
```

Comment:
Not too excited about this. What do you think about having either file or kube? And we can have cli option to that effect, change the yaml files. Less confusion.

Reply:
Oh ignore. Stupid comment. But please rename in to cliargs

Reply:
Not sure I agree that `cliargs` is a better name here. Why is `args` not sufficient?

Reply:
args also works.. just `in` seems a bit confusing..

Reply:
oof yeah, forgot that the name was `in` ... renamed to `args` :)